addons/event_crm/models/event_registration.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+
+
+class EventRegistration(models.Model):
+    _inherit = 'event.registration'
+
+    lead_ids = fields.Many2many('crm.lead', string='Leads', readonly=True, help="Link the leads generated from the registration.")
+
+    @api.model
+    def create(self, vals):
+        registration = super(EventRegistration, self).create(vals)
+        self.env['event.crm.rule'].search([])._create_leads(registration)
+        return registration
+
+    def write(self, vals):
+        registration = super(EventRegistration, self).write(vals)
+        for lead in self.lead_ids:
+            vals = {}
+            name = self.event_id.name + ' - '
+            if lead.from_rule_id.lead_creation_basis == 'order':
+                description = lead.from_rule_id._get_lead_description(lead.registration_ids)
+                vals.update({
+                    'description': description,
+                })
+                if self.partner_id and self.partner_id != self.env.ref('base.public_partner'):
+                    name += self.partner_id.name
+                    contact_name = self.partner_id.name
+                    email = self.partner_id.email
+                    phone = self.partner_id.phone
+                    mobile = self.partner_id.mobile
+                else:
+                    name += _('Anonymous')
+                    contact_name = email = phone = mobile = False
+            else:
+                name += self.name
+                contact_name = self.name
+                email = self.email
+                phone = self.phone
+                mobile = self.mobile
+            vals.update({
+                'name': name,
+                'contact_name': contact_name,
+                'email_from': email,
+                'phone': phone,
+                'mobile': mobile,
+            })
+            lead.write(vals)
+        return registration
+
+    def copy(self, default=None):
+        self.ensure_one()
+        vals = self.with_context(active_test=False).copy_data(default)[0]
+        vals = self._duplicate_registration(vals)
+        new_registration = self.with_context(lang=None).create(vals)
+        self.with_context(from_copy_translation=True).copy_translations(new_registration, excluded=default or ())
+        return new_registration
+
+    def _duplicate_registration(self, values):
+        values['main_registration_id'] = False
+        values['lead_ids'] = False
+        return values
```

Comment:
same here, return a dict and call it `vals.update(self._duplicate_registration())`